scram/main.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	
+	"syscall/js"
+	std_utils "globe-and-citizen/layer8/utils"
+)
+
+func main() {
+	close := make(chan struct{}, 0)
+
+	js.Global().Set("scram", map[string]interface{}{
+		"first": js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+			if len(args) != 4 {
+				panic(errors.New("all arguments are required: (password, iterationCount, salt, combinedNonce)"))
+			}
+
+			passwordArg := args[0].String()
+			iterationCountArg := args[1].Int()
+			saltArg := args[2].String()
+			combinedNonceArg := args[3].String()
+
+			if passwordArg == "" || iterationCountArg == 0 || saltArg == "" || combinedNonceArg == "" {
+				panic(errors.New("all arguments are required: (password, iterationCount, salt, combinedNonce)"))
+			}
+
+			salt, err := base64.StdEncoding.DecodeString(saltArg)
+			if err != nil {
+				panic(err)
+			}
+
+			saltedPassword := std_utils.HI(passwordArg, string(salt), iterationCountArg)
+			clientKey := std_utils.HmacSha256(saltedPassword, []byte("Client Key"))
+			storedKey := sha256.Sum256(clientKey)
+			serverKey := std_utils.HmacSha256(saltedPassword, []byte("Server Key"))
```

Comment:
Where does the string "Server Key" come from? Is this a public key?

Reply:
These could be any arbitrary messages so long as they are the same on the Client / Server